### ISSUE TITLE
Compiler optionInlineNone produces incorrect output

### ISSUE THREAD

**Issue by daniels220** [guillep/pharo-core#14323]
**Bug description**

The `optionInlineNone` compiler option is supposed to be equivalent to disabling all the other `optionInlineX` options, but in fact it is only sorta _half_-equivalent—it disables the inlining of the messages themselves but _not the blocks_, resulting in _wildly_ incorrect semantics:

* I discovered this with `and:`/`or:`, which are actually the least affected—they just lose their short-circuiting, both sides are always evaluated. But this is certainly still a problem, e.g. `a notNil and: [a < b]` suddenly isn't actually safe against a `nil DNU #<`.
* `ifTrue:ifFalse:` and `ifNil:ifNotNil:` just...always execute both branches—with the argument to the not-nil block set to nil, as well, in that case. They do return the value from the correct one, but are completely broken for control-flow purposes.
* `repeat`, `timesRepeat:` and `whileTrue[:]` only evaluate the loop body once.
* `to:[by:]do:` evaluates the body once, but with the "argument" set to nil rather than the starting value, almost invariably causing an error.
* I don't want to think about what `caseOf:[otherwise:]` does...probably execute all the cases?

**To Reproduce**

Basically, just compile anything with `<compilerOptions: #(+ optionInlineNone)>`. Even just a DoIt:

```
|i|
<compilerOptions: #(+ optionInlineNone)>
i:= 1.
10 timesRepeat: [i := i + 1].
i. " => yields 2, should be 11"
```

**Expected behavior**
`#(+ optionInlineNone)` should be exactly equivalent to `#(- optionInlineAndOr optionInlineCase optionInlineIf ...etc...)` and produce correct semantics.

**Possible solutions**
The question here is where to intervene. Right now the only method that produces the right answer when called in isolation is `RBMessageNode>>isInlined`—all the various `isInlineX` methods there can still answer true when `optionInlineNone` is set, and in turn, `RBBlockNode>>isInlined` also doesn't check it, which is the source of this bug. The simplest/shortest solution would be to just add a check at the top of `RBBlockNode>>isInlined` (**and** `isInlinedLoop`, which has independent senders), but this seems prone to future breakage—if for instance someone checks `isInlineAndOr` in some third place. As such, I think the _safest_ solution is to change all the `CompilationContext>>optionInlineX` methods to `^self optionInlineNone not and: [options includes: #optionInlineX]`. The specific check for `optionInlineNone` in RBMessageNode>>isInlined could then be removed. This way `optionInlineNone` is very literally equivalent to turning all the others off at a pretty low level, so it shouldn't matter how they're used.

**Expected development cost**
Pull request incoming shortly for the suggested approach above.

**Comment by daniels220** [comment]
...it looks like my suggested approach will (subtly!) break tons of compiler tests...but these compiler tests are also not testing what they need to be testing. They're setting up compiler options like `#(+ optionInlineAndOr optionInlineNone)`, clearly with the _intent_ of "inline `and:`/`or:` but nothing else", but in fact what they're getting is the broken behavior...but they aren't actually testing the ways it breaks! So uh. That's a thing.

The approach that gives the right sort of answer in this case is to make it so that `optionInlineNone` overrides the _default_ value of the other options, but they can still be included explicitly, but this is a bigger refactoring than I want to take on—right now that distinction between default and explicit is discarded pretty early on.

**Comment by guillep** [comment]
this is a great catch! thanks for looking into this!

about the tests, they may not be perfect but they exist at least :) do not hesitate to improve them

**Comment by daniels220** [comment]
Just took another look at it, and I misunderstood the meaning of the way the compiler options are set up—the tests are being run once with options like `#(+ optionInlineAndOr)` and once with `#(+ optionInlineNone)`. It's pretty straightforward to expand that to add a sanity-check with _no_ options (which I believe is un-optimized, it's just that the default in `Object class>>compiler` or wherever is to include most of the specific inlining optimizations), and most importantly a run with the problematic combination `#(+ optionInlineWhateverThisTestIsFor optionInlineNone)`. I did get to see this fail when I reverted the fix, then pass when I put it back, so that's encouraging! See #14397.
